src/Workspaces/Core/Portable/Rename/IRemoteRenamer.cs

```diff
@@ -208,4 +220,101 @@ internal class SerializableRenameLocations
         public SerializableRenameLocation[] StringsResult;
         public SerializableRenameLocation[] CommentsResult;
     }
+
+    internal class SerializableComplexifiedSpan
+    {
+        public TextSpan OriginalSpan;
+        public TextSpan NewSpan;
+        public ImmutableArray<(TextSpan oldSpan, TextSpan newSpan)> ModifiedSubSpans;
+
+        public ComplexifiedSpan Rehydrate()
+            => new ComplexifiedSpan(OriginalSpan, NewSpan, ModifiedSubSpans);
+
+        public static SerializableComplexifiedSpan Dehydrate(ComplexifiedSpan span)
+            => new SerializableComplexifiedSpan
+            {
+                OriginalSpan = span.OriginalSpan,
+                NewSpan = span.NewSpan,
+                ModifiedSubSpans = span.ModifiedSubSpans,
+            };
+    }
+
+    internal class SerializableRelatedLocation
+    {
+        public TextSpan ConflictCheckSpan;
+        public RelatedLocationType Type;
+        public bool IsReference;
+        public DocumentId DocumentId;
+        public TextSpan ComplexifiedTargetSpan;
+
+        public RelatedLocation Rehydrate()
+            => new RelatedLocation(ConflictCheckSpan, DocumentId, Type, IsReference, ComplexifiedTargetSpan);
+
+        public static SerializableRelatedLocation Dehydrate(RelatedLocation location)
+            => new SerializableRelatedLocation
+            {
+                ConflictCheckSpan = location.ConflictCheckSpan,
+                Type = location.Type,
+                IsReference = location.IsReference,
+                DocumentId = location.DocumentId,
+                ComplexifiedTargetSpan = location.ComplexifiedTargetSpan,
+            };
+    }
+
+    internal class SerializableConflictResolution
+    {
+        public string ErrorMessage;
+
+        public bool ReplacementTextValid;
+
+        public (DocumentId documentId, string newName) RenamedDocument;
+
+        public DocumentId[] DocumentIds;
+        public SerializableRelatedLocation[] RelatedLocations;
+        public ImmutableArray<(DocumentId, ImmutableArray<TextChange>)> DocumentTextChanges;
+        public (DocumentId, ImmutableArray<(TextSpan oldSpan, TextSpan newSpan)>)[] DocumentToModifiedSpansMap;
```

Comment:
Is there no serializable map type we can use? What's the benefit of array of tuples here?

Reply:
> Is there no serializable map type we can use? 

There is not unfortunately.  That's because this all goes through JSON.  And json can only represent maps using strings (and DocIds are not rountrippable to a simple string).  So this is effectively just serializing out all the key/values to tuples since that is something that can go through json.